### DISCUSSION TITLE
Installing troubles M1 MacOS "error: Not a valid ref: refs/remotes/origin/master" "fetch-pack: unexpected disconnect while reading sideband packet"

### DISCUSSION THREAD

**Discussion by clexp** [Homebrew/discussions#1259] [Getting started]
Trouble installing.  This user (me) is new to macOS - 1 week, New to Home brew - 2 days, bit of experience with Linux, typically windows end user, 

Issue : Engaging with denoobing exercises eg: using CLI package manager, compiling packages (within reason).  
`xcode-select --install`
seemed to work and not throw any errors
`curl -fsSL -o install.sh https://raw.githubusercontent.com/Homebrew/install/master/intall.sh`
seems to work.  
 had a look through the script and got a feel for it.  

So  ran the script.  Progressed well for a while, watched the verbose output, looked positive, but it ended with this:
`HEAD is now at bf0f38c72 Merge pull request #11083 from Bo98/ghp-bottle-tag
==> Tapping homebrew/core
remote: Enumerating objects: 99, done.
remote: Counting objects: 100% (99/99), done.
remote: Compressing objects: 100% (52/52), done.
fatal: the remote end hung up unexpectedly96.28 MiB | 7.04 MiB/s
fatal: early EOF
fatal: index-pack failed
Failed during: git fetch --force origin refs/heads/master:refs/remotes/origin/master`

and does not recognise any `brew` commands.  

I have googled "the remote end hung up unexpectedly" a bit and it seems I don't recognise any of the methods resolving this problem for other people (as in I don't think they apply here).  I don't have any firewall settings beyond what is delivered with the MacBook. My ISP is fine on my other machines.  I can still open the page hosting the shell script.  Not sure where to look.  

appreciate a little direction in problem solving.  Thanks

**Comment by carlocab** [comment]
Are you on an Intel or ARM Mac? Are you able to clone homebrew/core manually?

**Comment by clexp** [comment]
Hi,
Thanks for the really quick response and sorry for my slow reply.  I work shifts in healthcare and things are..  ..busy.  I am new to using GitHub and Mac so I need to read and trial cloning, and I will get back to you.  

This is an M1 Mac.  
Thanks

**Comment by clexp** [comment]
Thanks again for replying. 
I have cloned to my MacBook.  If I run the install.sh script I still get the same error,  cloning locally obviously does not mean the zsh will recognise `brew update` or any other are command.  Sorry to be really new at this.  Have you got a favourite link to point new starters at?
Thanks for your time and attention.

**Comment by clexp** [comment]
Hi,
I did an OS update and a few other things seemed to start working.  I tried `/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"`

`chris@Chriss-MacBook-Pro ~ % /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
==> Checking for `sudo` access (which may request your password).
Password:
==> This script will install:
/opt/homebrew/bin/brew
/opt/homebrew/share/doc/homebrew
/opt/homebrew/share/man/man1/brew.1
/opt/homebrew/share/zsh/site-functions/_brew
/opt/homebrew/etc/bash_completion.d/brew
/opt/homebrew
==> The Xcode Command Line Tools will be installed.

Press RETURN to continue or any other key to abort
==> /usr/bin/sudo /usr/sbin/chown -R chris:admin /opt/homebrew
==> Searching online for the Command Line Tools
==> /usr/bin/sudo /usr/bin/touch /tmp/.com.apple.dt.CommandLineTools.installondemand.in-progress
==> Installing Command Line Tools for Xcode-12.5
==> /usr/bin/sudo /usr/sbin/softwareupdate -i Command\ Line\ Tools\ for\ Xcode-12.5
Software Update Tool

Finding available software

Downloading Command Line Tools for Xcode
Downloaded Command Line Tools for Xcode
Installing Command Line Tools for Xcode
Done with Command Line Tools for Xcode
Done.
==> /usr/bin/sudo /bin/rm -f /tmp/.com.apple.dt.CommandLineTools.installondemand.in-progress
==> /usr/bin/sudo /usr/bin/xcode-select --switch /Library/Developer/CommandLineTools
==> Downloading and installing Homebrew...
remote: Enumerating objects: 934, done.
remote: Counting objects: 100% (568/568), done.
remote: Compressing objects: 100% (5/5), done.
remote: Total 934 (delta 563), reused 565 (delta 563), pack-reused 366
Receiving objects: 100% (934/934), 388.49 KiB | 2.49 MiB/s, done.
Resolving deltas: 100% (663/663), completed with 114 local objects.
From https://github.com/Homebrew/brew
 _+ de01cf170...88cf8681a SMillerDev-patch-1 -> origin/SMillerDev-patch-1  (forced update)
 -* [new branch]          dependabot/bundler/Library/Homebrew/sorbet-0.5.6386 -> origin/dependabot/bundler/Library/Homebrew/sorbet-0.5.6386
   ba578f4ae..94ce3236a  master             -> origin/master
 -* [new tag]             3.1.3              -> 3.1.3
HEAD is now at 94ce3236a Merge pull request #11263 from Bo98/upgrade-source
error: Not a valid ref: refs/remotes/origin/master
fatal: ambiguous argument 'refs/remotes/origin/master': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
error: 3828 bytes of body are still expected
fetch-pack: unexpected disconnect while reading sideband packet
fatal: early EOF
fatal: index-pack failed
Error: Fetching /opt/homebrew/Library/Taps/homebrew/homebrew-core failed!
fatal: invalid upstream 'origin/master'
Failed during: /opt/homebrew/bin/brew update --force --quiet
chris@Chriss-MacBook-Pro ~ %`

googling that first error gives me references to 'locked' files which I don't think is relevant here - the file/dir is missing.  Other results indicate I need a 'git update' type command though that may not be mac.  This is a new mac Any suggestions?

**Comment by clexp** [comment]
[This link]([url]https://stackoverflow.com/questions/60069083/homebrew-install-error-failed-during-git-fetch-origin-masterrefs-remotes-origi(url)) seems to have been really helpful for a number of people:
https://stackoverflow.com/questions/60069083/homebrew-install-error-failed-during-git-fetch-origin-masterrefs-remotes-origi

I tried all the steps in there, uninstalling and deleting the directories and re-installing after each step.  
`git config --global user.email yourgitemail@example.com`
`git config --global http.postBuffer 5242880000`
work well but
`git config --globle http.sslVerify "false"`
needed changing to "global"

during the install script now I get 
`"fetch-pack: unexpected disconnect while reading sideband packet"`

Then I tried increase the buffer:
`git config --global http.postBuffer 1048576000`

uninstalled with script, deleted the directory, then run the bash curl command again, still stuck here:
`error: 2611 bytes of body are still expected.27 MiB | 6.56 MiB/s
fetch-pack: unexpected disconnect while reading sideband packet
fatal: early EOF
fatal: index-pack failed
Failed during: git fetch --force origin refs/heads/master:refs/remotes/origin/master
chris@Chriss-MacBook-Pro /opt % `

===

So dead end there until I find more.  So I went back to suggestion to above to do it manually. 
'git clone https://github.com/Homebrew/install.git'
downloaded to a local dir nicely.  
'bash install.sh'
and installs up to the same point above. 

I am almost out of options.  
The last thing to do is to use the [desktop macOS app](https://desktop.github.com), but this is a download manager, and I do not think it will affect installing homebrew.  Once downloaded it will still be running the same script and falling apart with GitHub.  

Any pointers?

**Reply by carlocab** [reply]
Try this:
```
cd /opt && mkdir homebrew && curl -L https://github.com/Homebrew/brew/tarball/master | tar xz --strip 1 -C homebrew
echo 'eval "$(/opt/homebrew/bin/brew shellenv)"' >> ~/.zprofile
```
Then, restart your terminal and then do `brew update`.

**Comment by tlk** [comment]
Have you seen https://github.com/Homebrew/discussions/discussions/666?

**Comment by clexp** [comment]
Oh right, thanks for the rapid answer.  

chris@Chriss-MacBook-Pro /opt % mkdir homebrew
mkdir: homebrew: Permission denied
chris@Chriss-MacBook-Pro /opt % sudo mkdir homebrew
Password:
chris@Chriss-MacBook-Pro /opt % ls
homebrew
chris@Chriss-MacBook-Pro /opt % curl -L https://github.com/Homebrew/brew/tarball/master | tar xz --strip 1 -C homebrew
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   128  100   128    0     0    670      0 --:--:-- --:--:-- --:--:--   670
.dockerignore: Can't create '.dockerignore'
.editorconfig: Can't create '.editorconfig'
.github/: Can't create '.github'
.github/ISSUE_TEMPLATE.md: Failed to create dir '.github'
.github/ISSUE_TEMPLATE/: Failed to create dir '.github'
.github/ISSUE_TEMPLATE/bug.yml: Failed to create dir '.github'
.github/ISSUE_TEMPLATE/config.yml: Failed to create dir '.github'
.github/ISSUE_TEMPLATE/feature.yml: Failed to create dir '.github'
.github/PULL_REQUEST_TEMPLATE.md: Failed to create dir '.github'
.github/codecov.yml: Failed to create dir '.github'

Um.. is this a permissions issue?

chris@Chriss-MacBook-Pro /opt % sudo curl -L https://github.com/Homebrew/brew/tarball/master | tar xz --strip 1 -C homebrew
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   128  100   128    0     0    719      0 --:--:-- --:--:-- --:--:--   715
.dockerignore: Can't create '.dockerignore'
.editorconfig: Can't create '.editorconfig'
.github/: Can't create '.github'
.github/ISSUE_TEMPLATE.md: Failed to create dir '.github'
.github/ISSUE_TEMPLATE/: Failed to create dir '.github'
.github/ISSUE_TEMPLATE/bug.yml: Failed to create dir '.github'
.github/ISSUE_TEMPLATE/config.yml: Failed to create dir '.github'
.github/ISSUE_TEMPLATE/feature.yml: Failed to create dir '.github'
.github/PULL_REQUEST_TEMPLATE.md: Failed to create dir '.github'
.github/codecov.yml: Failed to create dir '.github'
.github/dependabot.yml: Failed to create dir '.github'
.github/workflows/: Failed to create dir '.github'


No don't think so.  

I went through issue 666, I too am on virgin media.  I switched off web safe, and with web safe off I got the above.  

Then I tried going to /opt and used `-o install.sh` flag as on [this](https://www.digitalocean.com/community/tutorials/how-to-install-and-use-homebrew-on-macos) tutorial.  Then I ran the script from the opt directory.  It downloaded and installed. I can now issue `brew` commands.  

SOLVED

Thanks for following this thread and your time in reading.  Sorry I had not found #666 and read through it.  I appreciate I have a lot of technical debt to address, that's why I am here.
